File: projecteuler/p10_Summation_of_primes.py
#encoding=utf-8
"""
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.
"""
#Runs in 0.706214 seconds
import time

startTime = time.clock()

def primesInRange(limit):
    potentialPrimesList = [True] * limit
    potentialPrimesList[0] = potentialPrimesList[1] = False
    for index, isPrime in enumerate(potentialPrimesList):
        if isPrime is True:
            yield index
            for x in xrange(index*2, limit, index):
                potentialPrimesList[x] = False
primeList = []
for prime in primesInRange(2000000):
    primeList.append(prime)

print sum(primeList), time.clock()-startTime

File: projecteuler-ruby/p10_summation_of_primes.rb
require 'set'

# The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

# Find the sum of all the primes below two million.

def primeNumbersUpTo(limit)
  nonPrimeSet = Set[]
  primeArr = [2]
  currentPrime = 2

  while currentPrime < limit do
    currentMultiple = 2
    while currentMultiple * currentPrime <= limit do
      nonPrimeSet.add(currentMultiple * currentPrime)
      currentMultiple += 1
    end
    nextPrime = currentPrime + 1
    while nonPrimeSet.include?(nextPrime) do
      nextPrime += 1
    end
    if nextPrime < limit
      primeArr << nextPrime
    end
    currentPrime = nextPrime
  end
  return primeArr
end

def sumOfAllPrimesBelow(num)
  primeNumbersUpTo(num).inject(:+)
end

puts sumOfAllPrimesBelow(2000000)
